RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+* It’s not for the “automatic” BC/FC support that can be done without any developer’s manual work (for example, the number of arguments mentioned in the Context). To apply versioning on the updated operator, the author needs to manually add a line in the version table and provide the resolution function for BC. This proposal is for BC/FC breakages that the automatic supports don’t apply. 
+* It does not include the BC/FC for package/deploy itself. The Python-only operators are transparent to TS and Edge clients, with the TS compilation. 
+
+
+# Proposal
+
+
+
+* **Eager changes**
+    * operator_versions.yaml** and **operator_upgraders.py** are added to register operator upgrades that are BC/FC breaking. **
+
+<p id="gdcalert1" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: undefined internal link (link text: "Example changes"). Did you generate a TOC? </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert2">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+
+[Example changes](#heading=h.enajdjh00ukn)
+        * Note: this will not cover functional operators
+        * The default value is zero
+    * **Use a single operator version number for all operators**
+        * This number may be shared by the deploy version, but separate from other file format versions
+    * **Older versions of operators registration are kept but they should only be matchable with special circumstances**
+    * **Newer version of the operator registry must specify an upgrader** that conforms to the older version of the operator schema. Its body is a TorchScript-able function that uses the newer version of operator to implement old semantics.
+    * [Improved BC testing] Tests that the old serialized version of the operator can still be loaded on the new runtime and run as expected need to be easy to add
+        * This seems straightforward for Torchscript and Edge testing but I’m not sure how it would work for deploy/package
+    * [Improved FC testing] Tests that the new version of the operator can still be loaded on old runtimes and run as expected need to be easy to add 
+        * This might require a new test job, which could be tricky to setup
+* **Torchscript changes**
+    * TorchScript Runtime build contains a table of operators and their corresponding versions
+    * Each serialized model contains a table of operators and their corresponding version according to the TorchScript compiler that generated the model
+        * Q: will this require a new serialization format?
+        * Yes, but it is easy to make it BC and FC
+    * During loading into the TorchScript compiler, TorchScript needs to match operator schema according to the table of operator versions stored in the package. This would generate IR that conforms to older schema. 
+    * Then a TorchScript pass takes IR of older schema and replaces older versions of operator invocations with bodies of upgraders.
```

Comment:
This implementation detail seems unnecessary -- torchscript can already run upgraders today

Reply:
@gmagogsfm, @tugsbayasgalan

Reply:
Yeah I think it would be easier if we just use upgrader as a builtin function just like how we do it now. What do you think @gmagogsfm ?

Reply:
Updated.